### ISSUE TITLE
fetch_token doesn't handle application-specific OAuth2 errors

### ISSUE THREAD

**Issue by bluekeyes** [oauthlib/oauthlib#431]
I ran into this problem working with the GitHub API. According to [their docs](https://developer.github.com/v3/oauth/#common-errors-for-the-access-token-request), GitHub uses several error types that are not part of the RFC and thus not included in `errors.py`. When `fetch_token` eventually calls `raise_from_error` via `validate_token_parameters`, no exception is raised because the error type is not found. As a result, you get a confusing validation error about a missing token:

```
oauthlib.oauth2.rfc6749.errors.MissingTokenError: (missing_token) Missing access token parameter.
```

I don't know if GitHub is actually following the spec here, but either way the result is confusing. I'm currently patching this in my application as follows:

``` python
# in oauth2/rfc6749/parameters.py

class ServerSpecificError(OAuth2Error):
    """The authorization server returned an error that is not part of the
    specification.
    """
    def __init__(self, error, **kwargs):
        self.error = error
        super().__init__(**kwargs)

def raise_from_error(error, params=None):
    kwargs = {
        'description': params.get('error_description'),
        'uri': params.get('error_uri'),
        'state': params.get('state')
    }
    for _, cls in inspect.getmembers(sys.modules[__name__], inspect.isclass):
        if cls.error == error:
            raise cls(**kwargs)
    raise ServerSpecificError(error, **kwargs)
```

I'm happy to submit this as a PR if it seems like an appropriate solution.

**Comment by sjamaan** [comment]
This would be helpful.

It would probably also be helpful if the http response code (or maybe the entire response body?) would be included, that if you're hitting a web page instead of a properly formatted JSON response, you get more to work with than just a bland "no token" exception.

**Comment by xmedeko** [comment]
+1 I would name it `UnknownServerError`.

**Comment by JonathanHuot** [comment]
cleaning/resurrecting old issues:

While I understand the issue, I have no clear scenario in mind. Could you draft a very quick sample to reproduce ?

**Comment by bluekeyes** [comment]
Sure, here's an example using `flask` and `requests-oauthlib`. I expect to see an error about [incorrect client credentials](https://developer.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/#incorrect-client-credentials) because I deliberately add `invalid` to the `client_secret` value before using it:

```py
import flask
import os

from flask import request
from flask.json import jsonify
from requests_oauthlib import OAuth2Session

app = flask.Flask(__name__)

client_id = '<id from github>'
client_secret = '<secret from github>'

auth_url = 'https://github.com/login/oauth/authorize'
token_url = 'https://github.com/login/oauth/access_token'

@app.route('/')
def login():
    github = OAuth2Session(client_id)
    url, state = github.authorization_url(auth_url)
    return flask.redirect(url)

@app.route('/callback', methods=['GET'])
def callback():
    github = OAuth2Session(client_id)
    token = github.fetch_token(token_url, client_secret=client_secret+'invalid', authorization_response=request.url)
    return jsonify({'status': 'success'})

if __name__ == '__main__':
    os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = "1"
    app.run(debug=True)
```

To reproduce the issue:

1. Register a new GitHub OAuth application with `http://localhost:5000/callback` as the callback URL
2. Put the client ID and client secret into the code above and run it
3. Visit `http://localhost:5000/` and complete the OAuth flow with GitHub

You should see the `MissingTokenError` instead of the actual response from GitHub. This makes debugging the issue pretty difficult, as it's not obvious that there's a typo in my client secret. I believe a similar thing will happen if the `redirect_uri` is incorrect, perhaps because one version uses HTTP while another uses HTTPS.

**Comment by JonathanHuot** [comment]
I have fixed it for `2.1.0` and added a catch-all `CustomOAuth2Error`, however for some reasons the `master` branch has broken the sample above.

I'll dig into it later

**Comment by JonathanHuot** [comment]
The `(Not Found)` error was due to a regression of the authorization-code flow (see https://github.com/oauthlib/oauthlib/issues/585) in requests-oauthlib when using github provider. That's is not related to this particular change.
